### PR TITLE
Now passing SPI chip selects through pinmux

### PR DESCRIPTION
**Pull request by HU90m**
*see commit messages*

### CODE REVIEW

File: rtl/system/sonata_pkg.sv [hunk 1]
```diff
@@ -6,10 +6,10 @@
 
 package sonata_pkg;
 
-  localparam int unsigned GPIO_NUM = 3;
   localparam int unsigned UART_NUM = 5;
-  localparam int unsigned I2C_NUM  = 2;
-  localparam int unsigned SPI_NUM  = 5;
+  localparam int unsigned I2C_NUM = 2;
+  localparam int unsigned SPI_NUM = 5;
+  localparam int unsigned GPIO_NUM = 3;
```
Comment (marnovandermaas) [inline rtl/system/sonata_pkg.sv:13]: Slightly amusing reordering as an artefact of your template change.

File: util/top_gen/generator.py [hunk 2]
```diff
@@ -364,8 +344,8 @@ def legacy_ordering(input_pin: OutputBlockIo) -> tuple[int, int, int]:
         ("rtl/templates/pinmux.sv.tpl", "rtl/system/pinmux.sv"),
         ("doc/ip/pinmux.md.tpl", "doc/ip/pinmux.md"),
     ):
-        print("Generating from template: " + template_file)
+        print("Generating from template:", template_file)
```
Comment (marnovandermaas) [inline util/top_gen/generator.py:348]: Is this removal of the space after the colon intentional?

File: util/top_gen/generator.py [hunk 3]
```diff
@@ -364,8 +344,8 @@ def legacy_ordering(input_pin: OutputBlockIo) -> tuple[int, int, int]:
...
         content = Template(filename=template_file).render(**template_variables)
-        Path(output_file).write_text(content)
+        Path(output_file).write_text(str(content))
```
Comment (marnovandermaas) [inline util/top_gen/generator.py:350]: Why was this necessary?

File: rtl/templates/sonata_pkg.sv.tpl [hunk 4]
```diff
@@ -6,12 +6,22 @@
...
   % endfor
 
+  // Width of block IO arrays
+  % for block in config.blocks:
+  % for b_io in block.ios:
+  % if b_io.length is not None:
+  localparam int unsigned ${f"{block.name}_{b_io.name}_WIDTH".upper()} = ${b_io.length};
+  % endif
+  % endfor
+  % endfor
```
Comment (marnovandermaas) [inline rtl/templates/sonata_pkg.sv.tpl:21]: Looks a bit complicated for just producing the GPIO width.

File: rtl/fpga/top_sonata.sv [hunk 5]
```diff
@@ -453,59 +430,46 @@ module top_sonata
...
+  assign ethmac_copi  = output_pins[OUT_PIN_ETHMAC_COPI];
+  assign lcd_cs       = output_pins[OUT_PIN_LCD_CS     ];
+  assign lcd_clk      = output_pins[OUT_PIN_LCD_CLK    ];
+  assign lcd_copi     = output_pins[OUT_PIN_LCD_COPI   ];
+  assign appspi_cs    = output_pins[OUT_PIN_APPSPI_CS  ];
+  assign appspi_clk   = output_pins[OUT_PIN_APPSPI_CLK ];
+  assign appspi_d0    = output_pins[OUT_PIN_APPSPI_D0  ];
+  assign rs232_tx     = output_pins[OUT_PIN_RS232_TX   ];
+  assign ser1_tx      = output_pins[OUT_PIN_SER1_TX    ];
+  assign ser0_tx      = output_pins[OUT_PIN_SER0_TX    ];
```
Comment (marnovandermaas) [inline rtl/fpga/top_sonata.sv:451]: Alignment off by one.

File: rtl/system/sonata_pkg.sv [hunk 6]
```diff
@@ -13,12 +13,13 @@ package sonata_pkg;
   localparam int unsigned GPIO_NUM = 3;
 
   // Width of block IO arrays
+  localparam int unsigned SPI_CS_WIDTH = 4;
   localparam int unsigned GPIO_IOS_WIDTH = 32;
   localparam int unsigned SPI_CS_NUM = 4;
```
Comment (marnovandermaas) [inline rtl/system/sonata_pkg.sv]: Are CS width and CS num the same?

**Review by marnovandermaas — COMMENTED** [review]
Nice work! Some comments from my side but in general I'm happy for this to go in if I2C, SPI and UART tests are all passing.

File: util/top_gen/generator.py (hunk 2 above)
Comment (HU90m) [reply util/top_gen/generator.py:348]: Yep, the default  value for python `print`'s `sep` argument is a space.

File: util/top_gen/generator.py (hunk 3 above)
Comment (HU90m) [reply util/top_gen/generator.py:350]: The `pyright` type checker will complains that `Template` could return types other than string. This makes it happy.

File: rtl/templates/sonata_pkg.sv.tpl (hunk 4 above)
Comment (HU90m) [reply rtl/templates/sonata_pkg.sv.tpl:21]: I don't think this is too complicated for what it is. It produces a width parameter for every block IO which has a length. So it can also be used for SPI CS.

I think this is a pretty clear way of expressing what it does

File: rtl/system/sonata_pkg.sv (hunk 6 above)
Comment (HU90m) [reply rtl/system/sonata_pkg.sv]: Yes! good spot this found it's way back in after rebasing. We can remove `SPI_CS_NUM` now.

File: util/top_gen/generator.py (hunk 3 above)
Comment (marnovandermaas) [reply util/top_gen/generator.py:350]: Cool well thanks for fixing this.

File: util/top_gen/generator.py (hunk 2 above)
Comment (marnovandermaas) [reply util/top_gen/generator.py:348]: Ok, great.

File: rtl/templates/sonata_pkg.sv.tpl (hunk 4 above)
Comment (marnovandermaas) [reply rtl/templates/sonata_pkg.sv.tpl:21]: Ok, it doesn't look wrong to me so let's keep it in.

File: rtl/fpga/top_sonata.sv (hunk 5 above)
Comment (marnovandermaas) [reply rtl/fpga/top_sonata.sv]: Now resolved.

File: rtl/bus/sonata_xbar_main.sv [hunk 7]
```diff
@@ -89,16 +91,16 @@ module sonata_xbar_main (
...
+    .tl_uart0_o        (tl_uart_o[0]),
+    .tl_uart0_i        (tl_uart_i[0]),
+    .tl_uart1_o        (tl_uart_o[1]),
+    .tl_uart1_i        (tl_uart_i[1]),
+    .tl_uart2_o        (tl_uart_o[2]),
+    .tl_uart2_i        (tl_uart_i[2]),
+    .tl_uart3_o        (tl_uart_o[3]),
+    .tl_uart3_i        (tl_uart_i[3]),
+    .tl_uart4_o        (tl_uart_o[4]),
+    .tl_uart4_i        (tl_uart_i[4]),
```
Comment (marnovandermaas) [inline rtl/bus/sonata_xbar_main.sv:103]: You added a space here.

Comment (marnovandermaas) [reply rtl/bus/sonata_xbar_main.sv:103]: No problem lets leave it.

File: rtl/system/sonata_pkg.sv (hunk 6 above)
Comment (marnovandermaas) [reply rtl/system/sonata_pkg.sv]: Now resolved.

File: vendor/lowrisc_ip/util/design/gen-otp-immutable-rom-ext-json.py
Comment (marnovandermaas) [inline vendor/lowrisc_ip/util/design/gen-otp-immutable-rom-ext-json.py:1]: What commit does this refer to? Also, is this necessary for this PR?

Comment (HU90m) [reply vendor/lowrisc_ip/util/design/gen-otp-immutable-rom-ext-json.py:1]: This pops up if you re-vendor `lowrisc-ip` in main

Comment (marnovandermaas) [reply vendor/lowrisc_ip/util/design/gen-otp-immutable-rom-ext-json.py:1]: Ok, I guess that's a change from upstream main.
